pubtools/_content_gateway/push_cgw.py

```diff
@@ -1,2 +1,196 @@
+        product_id = self.__get_product_id(product_name, product_code)
+        all_versions = self.cgw_client.get_versions(product_id)
+        self.pv_mapping = {(product_name, product_code, data.get('versionName')): (product_id, data.get('id')) for data
+                           in all_versions}
+
+    def __get_product_id(self, product_name, product_code):
+        product_id = self.product_mapping.get((product_name, product_code))
+        return product_id
+
+    def __get_version_id(self, product_name, product_code, version_name):
+        version_id = self.pv_mapping.get((product_name, product_code, version_name)) if self.pv_mapping else None
+        if version_id:
+            _, version_id = version_id
+        return version_id
+
+    def __get_file_id(self, file_item):
+        product_name = file_item.get('metadata').get('productName')
+        product_code = file_item.get('metadata').get('productCode')
+        version_name = file_item.get('metadata').get('productVersionName')
+
+        product_id = self.__get_product_id(product_name, product_code)
+        if not self.pv_mapping:
+            self.__create_product_version_mapping(product_name, product_code)
+        version_id = self.__get_version_id(product_name, product_code, version_name)
+        all_files = self.cgw_client.get_files(product_id, version_id)
+        file_mapping = {(product_name, product_code, version_name, data.get('downloadURL')): (
+            product_id, version_id, data.get('id')) for data in all_files}
+        pvf_id = file_mapping.get(
+            (product_name, product_code, version_name,
+             file_item.get('metadata').get('downloadURL'))) if file_mapping else (None, None, None)
+        return pvf_id
+
+    def process_product(self, item):
+        try:
+            product_name = item.get('metadata').get('name')
+            product_code = item.get('metadata').get('productCode')
+            LOG.debug(f"Fetching for product_id of product name- {product_name} and product code- {product_code}")
+            product_id = self.__get_product_id(product_name, product_code)
+
+            if item.get('state') == 'present':
+                if not product_id:
+                    LOG.info("No previous entries found for the given product name and product code")
+                    LOG.info("Creating product entry for the given product metadata")
+                    product_id = self.cgw_client.create_product(item.get('metadata'))
+                    # A new product is created so updating product mapping
+                    cgw_data = self.cgw_client.get_products()
+                    self.product_mapping = {(data['name'], data['productCode']): data['id'] for data in cgw_data}
+                    LOG.debug(f"Created a new product with product_id:- {product_id}")
+                    return
+                LOG.info(f"Product record found with product_id:- {product_id}")
+                item.get('metadata')['id'] = product_id
```

Comment:
It will never return any error since metadata will be a dict and I'm adding a new key-value to it.
To update any record the ids should be present in the request body which is not externally added by the user in the Yaml file. So, first I'm fetching the id and later adding it into the metadata for the record updation.

e.g:-
>>> x = {}
>>> x
{}
>>> x['metadata'] = {}
>>> x
{'metadata': {}}
>>> x['metadata']['data'] = "Test data"
>>> x
{'metadata': {'data': 'Test data'}}